pkg/rtc/mediatracksubscriptions.go

```diff
@@ -0,0 +1,499 @@
+		}
+	})
+	downTrack.OnPaddingSent(func(_ *sfu.DownTrack, size int) {
+		if t.params.Telemetry != nil {
+			t.params.Telemetry.OnDownstreamPacket(subscriberID, t.params.MediaTrack.ID(), size)
+		}
+	})
+	downTrack.OnRTCP(func(pkts []rtcp.Packet) {
+		if t.params.Telemetry != nil {
+			t.params.Telemetry.HandleRTCP(livekit.StreamType_DOWNSTREAM, subscriberID, t.params.MediaTrack.ID(), pkts)
+		}
+	})
+
+	downTrack.OnCloseHandler(func() {
+		go func() {
+			t.subscribedTracks.Delete(subscriberID)
+			if t.params.Telemetry != nil {
+				t.params.Telemetry.TrackUnsubscribed(context.Background(), subscriberID, t.params.MediaTrack.ToProto())
+			}
+
+			// ignore if the subscribing sub is not connected
+			if sub.SubscriberPC().ConnectionState() == webrtc.PeerConnectionStateClosed {
+				return
+			}
+
+			// if the source has been terminated, we'll need to terminate all of the subscribedtracks
+			// however, if the dest sub has disconnected, then we can skip
+			if sender == nil {
+				return
+			}
+			t.params.Logger.Debugw("removing peerconnection track",
+				"track", t.params.MediaTrack.ID(),
+				"subscriber", sub.Identity(),
+				"subscriberID", subscriberID,
+				"kind", t.params.MediaTrack.Kind(),
+			)
+			if err := sub.SubscriberPC().RemoveTrack(sender); err != nil {
+				if err == webrtc.ErrConnectionClosed {
+					// sub closing, can skip removing subscribedtracks
+					return
+				}
+				if _, ok := err.(*rtcerr.InvalidStateError); !ok {
+					// most of these are safe to ignore, since the track state might have already
+					// been set to Inactive
+					t.params.Logger.Debugw("could not remove remoteTrack from forwarder",
+						"error", err,
+						"subscriber", sub.Identity(),
+						"subscriberID", subscriberID,
+					)
+				}
+			}
+
+			t.NotifySubscriberMaxQuality(subscriberID, livekit.VideoQuality_OFF)
+			sub.RemoveSubscribedTrack(subTrack)
+			sub.Negotiate()
+		}()
+	})
+
+	t.subscribedTracks.Store(subscriberID, subTrack)
+	subTrack.SetPublisherMuted(t.params.MediaTrack.IsMuted())
+
+	// since sub will lock, run it in a goroutine to avoid deadlocks
+	go func() {
+		t.NotifySubscriberMaxQuality(subscriberID, livekit.VideoQuality_HIGH) // start with HIGH, let subscription change it later
+		sub.AddSubscribedTrack(subTrack)
+		sub.Negotiate()
+	}()
+
+	if t.params.Telemetry != nil {
+		t.params.Telemetry.TrackSubscribed(context.Background(), subscriberID, t.params.MediaTrack.ToProto())
+	}
+	return downTrack, nil
+}
+
+// RemoveSubscriber removes participant from subscription
+// stop all forwarders to the client
+func (t *MediaTrackSubscriptions) RemoveSubscriber(participantID livekit.ParticipantID) {
+	subTrack := t.getSubscribedTrack(participantID)
+	if subTrack != nil {
+		go subTrack.DownTrack().Close()
+	}
+}
+
+func (t *MediaTrackSubscriptions) RemoveAllSubscribers() {
+	t.params.Logger.Debugw("removing all subscribers", "track", t.params.MediaTrack.ID())
+
+	t.subscribedTracks.Range(func(_, val interface{}) bool {
```

Comment:
Keeping this a `sync.Map` and no locks surrounding it. But, got to admit I am a bit confused about how `Range` on a `sync.Map` works. It says a lock is not needed, but there is also some undefined behaviour kind of language when there are mutations during a `Range`. Cloud version uses a simple map with a lock. Maybe we can move to a simple map here also or just leave this as `sync.Map`. Please let me know if there are some gotchas here.

Reply:
You are right.. this would produce a race as Range doesn't guarantee that concurrent modifications will be reflected. (which makes sense)

Let's go back to a simple lock.

Reply:
Will do.